File: web3d/sim_worker.js
// Browser-side simulator: loads Pyodide and runs the Python 3D simulator
// (rocket3d / aero3d / guidance3d with Clarabel / sim3d) in this worker.
// Protocol (same data as the /api endpoints of main3d.py):
//   in:  {type: "input", keys, actions} | {type: "cfd", c_world}
//   out: {type: "progress", text} | {type: "state", json} | {type: "error", text}
const PYODIDE = "pyodide/";
const SIM_FILES = ["rocket3d.py", "aero3d.py", "guidance.py", "guidance3d.py", "guidance3d_legacy.py", "sim3d.py", "main3d.py"];
importScripts(PYODIDE + "pyodide.js");

let glue = null;
const queue = [];
self.onmessage = (ev) => { queue.push(ev.data); };

const progress = (text) => self.postMessage({ type: "progress", text });

async function main() {
  progress("正在加载 Python 运行时…");
  const py = await loadPyodide({ indexURL: PYODIDE });
  progress("正在加载 numpy / scipy / Clarabel…");
  await py.loadPackage(["numpy", "scipy", "clarabel"]);
  progress("正在加载仿真代码…");
  py.FS.mkdirTree("/home/pyodide/sim");
  const files = await Promise.all(SIM_FILES.map(async (f) => {
    const r = await fetch("py/" + f, { cache: "no-cache" });
    if (!r.ok) throw new Error(`py/${f}: HTTP ${r.status}`);
    return [f, await r.text()];
  }));
  for (const [f, src] of files) py.FS.writeFile("/home/pyodide/sim/" + f, src);
  const glueSrc = await (await fetch("sim_worker.py", { cache: "no-cache" })).text();
  const ns = py.globals.get("dict")();
  py.runPython(glueSrc, { globals: ns });
  glue = { advance: ns.get("advance"), state: ns.get("state"), input: ns.get("handle_input"), cfd: ns.get("handle_cfd") };
  progress("");
  let last = performance.now();
  let lastPost = 0;
  const tick = () => {
    try {
      while (queue.length) {
        const m = queue.shift();
        if (m.type === "input") glue.input(JSON.stringify({ keys: m.keys, actions: m.actions }));
        else if (m.type === "cfd") glue.cfd(JSON.stringify({ c_world: m.c_world }));
      }
      const now = performance.now();
      glue.advance((now - last) / 1000);
      last = now;
      if (now - lastPost >= 30) {
        lastPost = now;
        self.postMessage({ type: "state", json: glue.state() });
      }
    } catch (e) {
      self.postMessage({ type: "error", text: String(e && e.message || e) });
      return;
    }
    setTimeout(tick, 2);
  };
  tick();
}

main().catch((e) => self.postMessage({ type: "error", text: String(e && e.message || e) }));

File: web3d/sim_worker.py
"""Browser-side glue: runs the 3D simulator inside Pyodide (Web Worker).

Same physics/guidance/state as ``python main3d.py``; the real-time loop is
main3d.Runner's, driven by the worker's timer instead of a thread.
"""

import json
import sys

sys.path.insert(0, "/home/pyodide/sim")

import main3d
from sim3d import PHYSICS_DT, Sim3D

sim = Sim3D(cfd=True)
_acc = 0.0


def advance(real: float) -> None:
    """Advance by ``real`` wall seconds (x time warp); never fast-forwards to
    catch up after a slow step, like main3d.Runner."""
    global _acc
    _acc += min(0.1, max(0.0, real)) * sim.time_warp
    _acc = min(_acc, 0.1 * max(1.0, sim.time_warp))
    n = 0
    while _acc >= PHYSICS_DT and n < 2400:
        sim.step(PHYSICS_DT)
        _acc -= PHYSICS_DT
        n += 1
    if n >= 2400:
        _acc = 0.0


def state() -> str:
    return json.dumps(main3d._finite(main3d.snapshot(sim)), separators=(",", ":"), allow_nan=False,
                      default=lambda o: main3d._finite(float(o)) if hasattr(o, "__float__") else str(o))


def handle_input(payload: str) -> None:
    p = json.loads(payload)
    with sim.lock:
        sim.keys = {k: bool(v) for k, v in p.get("keys", {}).items()}
        for action in p.get("actions", []):
            main3d.apply_action(sim, action)


def handle_cfd(payload: str) -> None:
    p = json.loads(payload)
    if sim.cfd is not None and isinstance(p.get("c_world"), list):
        sim.cfd.update(p["c_world"])
